test/test_torch.py

```diff
@@ -7661,6 +7661,45 @@ def test_copy_many_to_one(self):
         # storage to a single storage would cause RuntimeError to be thrown
         self.assertRaises(RuntimeError, lambda: torch.zeros(1, 6).expand(5, 6).copy_(torch.zeros(5, 6)))
 
+    def test_copy_float16(self):
+        # Check that fbgemm code no longer reads memory out of bounds, see
+        # copy_impl and fbgemm::Float16ToFloat_ref.
+        # https://github.com/pytorch/pytorch/issues/88543
+
+        # Types to test different code paths in copy_impl.
+        dtypes = (
+            # out_dtype, src_dtype, is_fbgemm
+            (torch.float32, torch.float16, True),  # fbgemm
+            (torch.float16, torch.float32, True),  # fbgemm
+            (torch.float32, torch.float32, False),  # TensorIterator
+        )
+
+        # TODO: fbgemm and TensorIterator behave differently.
+        cases = (
+            # out_shape, src_shape, ok_fbgemm, ok_ti
+            # These cases used to crash with fbgemm, make sure these also raise
+            # exceptions with TensorIterator.
+            ((1, 2, 3), (0, 2, 3), True, False),  # same strides, not allowed by TI
+            ((1, 5, 6), (4, 5, 6), False, False),  # same strides, not allowed by TI
+            (1, (0, 2, 3), True, False),  # different strides
+            ((4, 5, 6), (0, 2, 3), True, False),  # different strides
+            ((4, 5, 6), (1, 2, 3), True, False),  # different strides
+
+            # These cases should pass with fbgemm and TensorIterator.
+            ((4, 5, 6), (1, 5, 6), True, True),  # same strides
+            ((4, 5, 6), (4, 5, 6), True, True),  # same strides
+            ((0, 2, 3), 1, True, True),  # different strides, allowed by TI
+            ((4, 5, 6), (4, 5, 1), True, True),  # different strides, allowed by TI
+        )
+
+        for (out_shape, src_shape, ok_fbgemm, ok_ti), (out_dtype, src_dtype, is_fbgemm) in itertools.product(cases, dtypes):
+            out = torch.ones(out_shape, dtype=out_dtype)
+            src = torch.ones(src_shape, dtype=src_dtype)
+            if (ok_fbgemm and is_fbgemm) or (ok_ti and not is_fbgemm):
+                out.copy_(src)
```

Comment:
Probably a good "source of truth" would be to run the same operation on CUDA and compare the results.